src/cli.ts

```diff
@@ -25,35 +40,103 @@ async function main(argv: minimist.ParsedArgs) {
   }
 
   const argIn: string = argv._[0] || argv.input
-  const argOut: string = argv._[1] || argv.output
+  const argOut: string | undefined = argv._[1] || argv.output // the output can be omitted so this can be undefined
+
+  const ISGLOB = isGlob(argIn)
+  const ISDIR = isDir(argIn)
+
+  if ((ISGLOB || ISDIR) && argOut && argOut.includes('.d.ts')) {
+    throw new ReferenceError(
+      `You have specified a single file ${argOut} output for a multi file input ${argIn}. This feature is not yet supported, refer to issue #272 (https://github.com/bcherny/json-schema-to-typescript/issues/272)`
+    )
+  }
 
   try {
-    const schema: JSONSchema4 = JSON.parse(await readInput(argIn))
-    const ts = await compile(schema, argIn, argv as Partial<Options>)
-    await writeOutput(ts, argOut)
+    // Process input as either glob, directory, or single file
+    if (ISGLOB) {
+      await processGlob(argIn, argOut, argv as Partial<Options>)
+    } else if (ISDIR) {
+      await processDir(argIn, argOut, argv as Partial<Options>)
+    } else {
+      await processFile(argIn, argOut, argv as Partial<Options>)
+    }
   } catch (e) {
     console.error(whiteBright.bgRedBright('error'), e)
     process.exit(1)
   }
 }
 
-function readInput(argIn?: string) {
-  if (!argIn) {
-    return new Promise(stdin)
+// check if path is an existing directory
+function isDir(path: string): boolean {
+  return existsSync(path) && lstatSync(path).isDirectory()
+}
+
+async function processGlob(argIn: string, argOut: string | undefined, argv: Partial<Options>) {
+  const files = await glob(argIn) // execute glob pattern match
+
+  if (files.length === 0) {
+    throw ReferenceError(
+      `You passed a glob pattern "${argIn}", but there are no files that match that pattern in ${process.cwd()}`
+    )
   }
-  return readFile(resolve(process.cwd(), argIn), 'utf-8')
+  // create output directory if it does not exist
+  if (argOut && !existsSync(argOut)) {
+    await mkdirp(argOut)
+  }
+
+  Promise.all(
+    files.map(file => {
+      const outPath = argOut && `${argOut}/${basename(file, '.json')}.d.ts`
+      processFile(file, outPath, argv)
+    })
+  )
 }
 
-function writeOutput(ts: string, argOut: string): Promise<void> {
+async function processDir(argIn: string, argOut: string | undefined, argv: Partial<Options>): Promise<void[]> {
+  const files = getPaths(argIn)
+
+  return Promise.all(
+    files.map(file => {
+      if (!argOut) {
+        processFile(file, argOut, argv)
+      } else {
+        let outPath = pathTransform(argOut, file)
+        if (!isDir(dirname(outPath))) {
+          _mkdirp.sync(dirname(outPath))
+        }
+        outPath = outPath.replace('.json', '.d.ts')
+        processFile(file, outPath, argv)
```

Comment:
And here

Reply:
What change are you expecting here?

Reply:
It’s been a while since I’ve touched JS concurrency, but are promises actually executed concurrently, with the Promise.all revolving when the longest promise resolves, if you don’t return them in all these places?

Reply:
I sadly do not know. Async/await behavior and promises are confusing for me so I don't know exactly what would happen. Assuming the mdn docs are true [Promise.all](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all) does resolve when _all_ promises resolve.